### ISSUE TITLE
[BUG] Custom proxy module for vcenter/esxi doesn't behave correctly

### ISSUE THREAD

**Issue by Dejv56** [saltstack/salt#61454]
**Description**
I am trying to create custom proxy module for vcenter/esxi, but i am having issues with reopening/maintaining connection with vcenter/esxi via salt.utils.vmware.get_service_instance() or using directly pyVim.connect.SmartConnect(). Connection always fails after few minutes/hours/days and then running any function on salt-proxy results in "**ERROR: Not enough permissions. Required privilege: System.View**". Only way to fix it is restarting salt-proxy service.

Another strange behauviour is when i start/restart salt-proxy it shows ESTABILISTED via "lsof -i@<vsphere/esx ip>" no matter the "PROXY_ALWAYS_ALIVE" or "PROXY_KEEP_ALIVE" option. After running some function on salt-proxy, connection always drops and it is created only to run requested function.

When i try the same using bare Python it works just fine, connection >> ok >> timeout with error above >> disconnect >> connection >> ok.

I used https://docs.saltproject.io/en/latest/topics/proxyminion/index.html when creating proxy module and used already created modules as reference.

**Setup**
pillar/proxy/esx.sls
```
proxy:
  proxytype: esx
  host: <ip>
  username: <user>
  password: <pass>
```

salt/_modules/esx.py
```
from __future__ import absolute_import, print_function, unicode_literals
import salt.utils.platform

__virtualname__ = 'esx'

def __virtual__():
    '''
    Check if:
      1) Is target minion proxy?
      2) Is proxy minion of correct proxytype?
    '''
    if not salt.utils.platform.is_proxy():
        return False, 'Module must be run on a proxy minion.'
    if __opts__.get('proxy', {}).get('proxytype') != __virtualname__:
        return False, 'Proxytype does not match: {}'.format(__virtualname__)
    return True

def list_vms():
    return __proxy__['esx.list_vms']()
```
salt/_proxy/esx.py
```
from __future__ import absolute_import, print_function, unicode_literals

# Import python libs
from pyVim.connect import SmartConnect, Disconnect
from pyVmomi import vim
import salt.utils.vmware
import logging
import ssl
from salt.exceptions import SaltException

# This must be present or the Salt loader won't load this module
__proxyenabled__ = ['esx']

DETAILS = {}

# Set up logging
log = logging.getLogger(__name__)

def init(opts):
    proxy_dict = opts.get('proxy', {})
    log.info('esx proxy trying connection...')
    try:
        ctx = ssl._create_unverified_context()
        DETAILS['server'] = SmartConnect(
            host=proxy_dict['host'],
            user=proxy_dict['username'],
            pwd=proxy_dict['password'],
            sslContext=ctx
            )   
        log.info('esx proxy connection successful')
        DETAILS["initialized"] = True
        return True
    except Exception as e:
        DETAILS["initialized"] = False
        log.info('esx proxy connection NOT successful')
        log.error(e)
        return False


def alive(opts):
    return ping()
    
def initialized():
    return DETAILS.get("initialized", False)

def grains():
    return {}

def ping():
    log.info('esx proxy pinging...')
    try:
        list_vms()
        log.info('esx proxy ping successful')
        return True
    except Exception as e:
        log.info('esx proxy ping NOT successful')
        log.error(e)
        return False

def shutdown(opts):
    log.info('esx proxy disconnecting')
    try:
        Disconnect(DETAILS['server'])
        log.info('esx proxy disconnected')
        return True
    except Exception as e:
        log.error('Cannot close connection, check error: \n{}'.format(e))
        return False

def list_vms():
    '''
    List all VMs present on ESX host.
    '''
    ret = []
    vm_properties = ['name']

    vm_list = salt.utils.vmware.get_mors_with_properties(
        DETAILS['server'], vim.VirtualMachine, vm_properties
    )

    for vm in vm_list:
        ret.append(vm['name'])

    return sorted(ret)
```

**Versions Report**
```
Salt Version:
          Salt: 3002.2

Dependency Versions:
          cffi: 1.13.2
      cherrypy: unknown
      dateutil: 2.8.2
     docker-py: 4.2.0
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 3.0.3
       libgit2: 0.28.4
      M2Crypto: 0.35.2
          Mako: Not Installed
       msgpack: 0.5.6
  msgpack-pure: Not Installed
  mysql-python: 1.4.6
     pycparser: 2.17
      pycrypto: Not Installed
  pycryptodome: 3.12.0
        pygit2: 0.28.2
        Python: 3.6.10 (default, Dec 19 2019, 15:48:40) [GCC]
  python-gnupg: Not Installed
        PyYAML: 5.1.2
         PyZMQ: 17.0.0
         smmap: Not Installed
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.2.3

System Versions:
          dist: sled 15.2
        locale: UTF-8
       machine: x86_64
       release: 5.3.18-24.15-default
        system: Linux
       version: SLED 15.2
```
</details>

**Comment by garethgreenaway** [comment]
@Dejv56 Thanks for the report. Based on the output above it sounds like the connection to ESXI is getting terminated and the proxy minion is not reconnecting.

**Comment by garethgreenaway** [comment]
Closing due to lack of activity.

**Comment by Dejv56** [comment]
I don't really undestand, why this issue got closed. As i see it, the proxy minion feature is not working as intended, or it is working differently than it is described in documentation.

**Comment by dmurphy18** [comment]
@Dejv56 The Salt core support for VMware is being moved to the VMware Salt Extensions, given that when they were written, pyvmomi was circa 5.x and not really functional now with the latest v8.0.0.x.
We have made a list of the functions and are busy porting them to Salt VMware Extensions, starting with the ESXi ones since they are most in demand.  Cannot give a timeframe as to when finished, but actively working them now. Reason I was surfing through old issues to see problems, when I noticed you needed a better answer.

With them in Salt VMware Extensions, can release independent of Salt, hence hope to get functionality into users hands quicker, here is the link to the GitHub repository for Salt VMware Extensions https://github.com/saltstack/salt-ext-modules-vmware
